CyberBoard 64-bit process

wsu-cb opened the pull request:
Dale,

This pull request makes CB able to build and run as either a 32-bit or 64-bit program.   This includes file format changes that further extend the previous v4.00 (`/id32`) format to support 64-bit `size_t` values.

This is a draft pull request because I would like your assessment of the following issues:

Since there hasn't been an actual release with 32bit ids in the file format, I have left the "new" file format as still being v4.00 even though it has changed again to support 64bit `size_t`.  This is my understanding of the conclusion of previous discussions, but I want to be sure this is still the desired behavior.

The default behavior of CB, in both 32- and 64-bit, is still to use the v3.90 file format.  I changed the command line switch to enable v4.00 to `/id32cb64` to make clear that the new v4.00 files have both changes.

I have used a minor compression technique on the 32bit id values (19a041a2acb27eff9d39a620023483dee05d6cf0) as well as Invalid_v<size_t> (f5bf4668dc9755c60ed76c70536e365a79be952b).  This (typically) results in files that are actually smaller than v3,90, but does make the file format less intuitive.  The reduction in size is quite small; it might not be worth the extra complexity.

GShr/CyberBoard.h disables warnings 4701 and 4703 at line 80.  They are uninitialized variable warnings.  I fixed some uninitialized vars at 1f65b6eaaafe4d0c95a988f9828d51f23751fa85.  However, I wasn't sure what to do with the rest, which is why I disabled the warnings.  I hope you can un-disable the warnings, and figure out some good initial values for the remaining vars.

While the process now supports 64bit, and some variables are actually 64bit, not everything that might have been widened actually was.  For example, I haven't tried to support individual bitmaps >=2G.  (I don't even know whether Windows supports that.)  Please let me know if there is anything specific that you think should be widened.

Resolve #13

DLLarson commented:
Hi Bill,

First off, it's good to hear you're healthy and still beating on code! (Anything can happen these days. ;) )

Just a warning that It's a large PR so it'll take me into next week to get through it.

-Dale

DLLarson commented:
Hi Bill,

I'm still trying to get at this but the holidays with family in-bound has been cutting into the quiet time I need to focus on code.

I've downloaded, built all models with debug and release and ran a x64 mode version. It's really cool to see CB show up in 64 bit mode looking like it always has (but of course it should!). But I realize it took real effort to get it there so that colors my view. 

So I'm asking for some patience on this one (not that you've made any demands as such--just wanted to let you know.)

This is a really great evolution of the code given its humble Win16 real mode beginnings.

-Dale

DLLarson commented:
Hi Bill,

Sorry for the long holiday hiatus. I'll be getting to this PR this week.

-Dale

DLLarson commented:
>Since there hasn't been an actual release with 32bit ids in the file format, I have left the "new" file format as still being v4.00 even though it has changed again to support 64bit size_t. This is my understanding of the conclusion of previous discussions, but I want to be sure this is still the desired behavior.

Yep. Modify to your heart's content until an official style release. Very low risk here since other people don't seem to be using the code.

>I have used a minor compression technique on the 32bit id values (19a041a) as well as Invalid_v<size_t> (f5bf466). This (typically) results in files that are actually smaller than v3,90, but does make the file format less intuitive. The reduction in size is quite small; it might not be worth the extra complexity.

Complexity added without a clear advantage should probably be omitted.

-Dale

DLLarson commented:
>GShr/CyberBoard.h disables warnings 4701 and 4703 at line 80. They are uninitialized variable warnings. I fixed some uninitialized vars at 1f65b6e. However, I wasn't sure what to do with the rest, which is why I disabled the warnings. I hope you can un-disable the warnings, and figure out some good initial values for the remaining vars.

For the CBDesign program with 4701 and 4703 enabled:

File `Rotate.cpp` Lines 145 and 146 can be set to zero. They will always be updated by the following loop so this should appease the compiler. (Note that Github won't allow me to add this sort of comment in the inspection feature if the code isn't part of a diff window.)

For the CBDesign program with 4701 and 4703 enabled:

The same code as above tripped the warning so fixing the above should cover all cases unless I'm missing something.

-Dale

DLLarson commented:
My initial sweep through the code is done. I'm going to spend some time testing the actual program against legacy files etc. over the next few days.

My one general comment is that it seems like when we change a standard C/C++ data type such as `int` to another more specific type like a word or byte we should use a standard C/C++ type such as `uint16_t` or `uint8_t` rather than using the MSoft equivalents like `WORD` or `BYTE`. This might not make sense when calling underlying Win32 APIs but as an overall principle we would end up with more portable code. I've tagged a few examples of these in the inspection comments.

-Dale

wsu-cb commented:
> My one general comment is that it seems like when we change a standard C/C++ data type such as `int` to another more specific type like a word or byte we should use a standard C/C++ type such as `uint16_t` or `uint8_t` rather than using the MSoft equivalents like `WORD` or `BYTE`. This might not make sense when calling underlying Win32 APIs but as an overall principle we would end up with more portable code. I've tagged a few examples of these in the inspection comments.
> 
> -Dale

I thought I was already doing so.  If I _added_ uses of `WORD` or `BYTE`, then it was carelessness, and should indeed be fixed.  I have been more inconsistent about _replacing_ `BYTE`/`WORD`/`DWORD` when I

wsu-cb commented:
ARGH!  My new computer did something weird and submitted the preceding comment, AND CLOSED THIS PR!  That was completely unintentional!

wsu-cb commented:
I was trying to say:

I thought I was already doing so.  If I _added_ uses of `BYTE`/`WORD`/`DWORD`, then it was carelessness, and absolutely should be fixed.  I am more inconsistent about _replacing_ `BYTE`/`WORD`/`DWORD`.  In those cases, I sometimes retain them if I think it makes diff comparisons easier to understand, but a lot of it is just how I feel at the time.

wsu-cb commented:
> > Since there hasn't been an actual release with 32bit ids in the file format, I have left the "new" file format as still being v4.00 even though it has changed again to support 64bit size_t. This is my understanding of the conclusion of previous discussions, but I want to be sure this is still the desired behavior.
> 
> Yep. Modify to your heart's content until an official style release. Very low risk here since other people don't seem to be using the code.
> 
> > I have used a minor compression technique on the 32bit id values ([19a041a](https://github.com/CyberBoardPBEM/cbwindows/commit/19a041a2acb27eff9d39a620023483dee05d6cf0)) as well as Invalid_v<size_t> ([f5bf466](https://github.com/CyberBoardPBEM/cbwindows/commit/f5bf4668dc9755c60ed76c70536e365a79be952b)). This (typically) results in files that are actually smaller than v3,90, but does make the file format less intuitive. The reduction in size is quite small; it might not be worth the extra complexity.
> 
> Complexity added without a clear advantage should probably be omitted.
> 
> -Dale

Do you want to wait to do more testing until after I remove the file compression?  By not changing the file format at each step, it means that any v4.00 files you save now will be unreadable once I remove this compression since both formats will be called v4.00.

DLLarson commented:
>Do you want to wait to do more testing until after I remove the file compression? By not changing the file format at each step, it means that any v4.00 files you save now will be unreadable once I remove this compression since both formats will be called v4.00.

Good point! I think I should wait until you commit that change so as not to have to redo the testing again.

DLLarson commented:
> I sometimes retain them if I think it makes diff comparisons easier to understand, but a lot of it is just how I feel at the time.

Understood. I'm leaving it up to you. I just thought I'd raise the point.

DLLarson commented:
>ARGH! My new computer did something weird and submitted the preceding comment, AND CLOSED THIS PR!

I guess the good news there is you have a *new* computer! ;)

DLLarson commented:
Hi Bill,

>>Do you want to wait to do more testing until after I remove the file compression? 

>Good point! I think I should wait until you commit that change so as not to have to redo the testing again.

Did these commits also remove the compression code?

-Dale

wsu-cb commented:
> Did these commits also remove the compression code?
> 
> -Dale

Yes, these commits removed the compression (I rebased the code to make the compression disappear from the history), but I have more bug-fixing to do.  That's why I didn't request another review yet.

DLLarson commented:
Ok... Thanks!

wsu-cb commented:
Dale,

I think I am ready for a serious review now.  I have removed the compression logic, fixed the rest of the uninitialized vars, fixed various other bugs, and replaced some of the Windows types with stdint int types.  Unfortunately, I discovered that Win32 `DWORD_PTR` is not defined consistently with `uintpr_t`, so `virtual void CWnd::WinHelp(DWORD_PTR dwData, UINT nCmd)` doesn't get overridden properly if the override uses `uintptr_t` instead of `DWORD_PTR`.  (In Win64, `uintpr_t` can substitute for DWORD_PTR.)  Therefore, the GM and  GP `CMainframe::WinHelp() ` functions must be declared using `DWORD_PTR`.

DLLarson commented on GShr/CellForm.cpp:
```diff
@@ -172,15 +172,15 @@ void CCellForm::CreateHexMask()
 }
 
-void CCellForm::FindCell(int x, int y, int& row, int& col)
+void CCellForm::FindCell(LONG x, LONG y, CB::ssize_t& row, CB::ssize_t& col)
```
BUMP.

DLLarson reviewed (approved):
I had one comment about LONG that was passed over. But it's a little thing so I'm approving the code inspection. I still need to run a variety of legacy game, gamebox and scenario files through the programs to sanity check the serializations.

This is a huge commit reflecting the gargantuan task you undertook. Hopefully this will get the platform target mods out of the way and make future commits easier to digest.

Really nice work!
-Dale

DLLarson commented:
Hi Bill,

I want to clarify this statement:

>The default behavior of CB, in both 32- and 64-bit, is still to use the v3.90 file format. I changed the command line switch to enable v4.00 to /id32cb64 to make clear that the new v4.00 files have both changes.

So if I don't use the switch I get files that are compatible with the current released version of CB and if I use the switch I get files that are upgraded from the shipped version?

An aside...

Testing the program file loads and saves is quite laborious so I'm exploring driving the process using `pywinauto` to automate testing for all the files in a directory tree (for instance.)

-Dale

wsu-cb commented:
> Hi Bill,
> 
> I want to clarify this statement:
> 
> > The default behavior of CB, in both 32- and 64-bit, is still to use the v3.90 file format. I changed the command line switch to enable v4.00 to /id32cb64 to make clear that the new v4.00 files have both changes.
> 
> So if I don't use the switch I get files that are compatible with the current released version of CB and if I use the switch I get files that are upgraded from the shipped version?
> 
> -Dale

Before my changes (tag v3.5-prerelease-2), the file version is 3.90.  With my changes, but no switch, the file version is still 3.90, and should be compatible with files from before my changes.  With my changes and the switch, the file version is 4.00, and requires my changes to handle those files.

HOWEVER, my understanding is that there has been no released CB that supports v 3.90.  To be more precise, my understanding is that the latest release is 3.10 from Jan 2010, and that is using file v3.10.  Therefore, even if you don't use the switch, files are NOT compatible with the current released CB.

DLLarson commented:
>HOWEVER, my understanding is that there has been no released CB that supports v 3.90. To be more precise, my understanding is that the latest release is 3.10 from Jan 2010, and that is using file v3.10. Therefore, even if you don't use the switch, files are NOT compatible with the current released CB.

Ok. This is part my brain didn't catch on to. So V3.90, in the end, will just be a specter version as no official files will ever be created with it. V4.00 will be the true file version of the first shipped version. That pretty much makes the switch we use superfluous in the bigger picture as all my testing should really focus on loading compatibility with 3.10 and earlier with only saves to the V4.00 format. After this commit we will have the ability to run either the 32 bit or x64 versions with file compatibility between both platform builds.

Does this capture things correctly?

-Dale

wsu-cb commented:
> Ok. This is part my brain didn't catch on to. So V3.90, in the end, will just be a specter version as no official files will ever be created with it. V4.00 will be the true file version of the first shipped version. That pretty much makes the switch we use superfluous in the bigger picture as all my testing should really focus on loading compatibility with 3.10 and earlier with only saves to the V4.00 format. After this commit we will have the ability to run either the 32 bit or x64 versions with file compatibility between both platform builds.
> 
> Does this capture things correctly?
> 
> -Dale

It kind of looks that way.  For a while, I thought there might be a CB 3.5 that would be (tag v3.5-prerelease-2 + bug fixes), which would have had file v3.90, but it hasn't happened, and at this point I infer that it probably won't.

DLLarson commented:
>It kind of looks that way. For a while, I thought there might be a CB 3.5 that would be (tag v3.5-prerelease-2 + bug fixes), which would have had file v3.90, but it hasn't happened, and at this point I infer that it probably won't.

Perhaps the switch should be reversed or removed?

-Dale

wsu-cb commented:
> > It kind of looks that way. For a while, I thought there might be a CB 3.5 that would be (tag v3.5-prerelease-2 + bug fixes), which would have had file v3.90, but it hasn't happened, and at this point I infer that it probably won't.
> 
> Perhaps the switch should be reversed or removed?
> 
> -Dale

Files with large ID values or rotated geoboard units can't be saved in file format v3.90, but most files still can, so I would be inclined to keep the switch, but I could see reversing it.  (Files that can't be saved in v3.90 format throw exceptions when saving.)

I am inclined to actually go farther:  I searched through the code, and I can't see how file v3.90 is different from file v3.10.  Am I missing something?  If they really are the same, why don't we make the switch actually produce file v3.10 output?  We can still read files that are marked v3.90 (since I at least still have a bunch of v3.90 files), but change the code to only write files as either v3.10 or v4.00.

DLLarson commented:
>If they really are the same, why don't we make the switch actually produce file v3.10 output? We can still read files that are marked v3.90 (since I at least still have a bunch of v3.90 files), but change the code to only write files as either v3.10 or v4.00.

I would go even farther...Dispense with the switch and only save v4.0 format but support loads for the earlier versions. That's what is normally done when the file formats are evolved. I just don't see the point of having a switch that saves as the shipped version. Wouldn't it imply the program can downgrade from 4.0 to 3.10?

wsu-cb commented:
> > If they really are the same, why don't we make the switch actually produce file v3.10 output? We can still read files that are marked v3.90 (since I at least still have a bunch of v3.90 files), but change the code to only write files as either v3.10 or v4.00.
> 
> I would go even further...Dispense with the switch and only save v4.0 format--but support loads for the earlier versions. That's what is normally done when the file formats are evolved. I just don't see the point of having a switch that saves as the shipped version. Wouldn't it imply the program can downgrade from 4.0 to 3.10?

The point of the switch is that uptake of the new version has been slow.  I hope more people would be willing to run CB 3.5/4.0 Alpha/Beta/Release if they could still interoperate with other users who were still using CB 3.1.  However, that's dependent on how restrictive (in the sense of being incompatible with new features) using file v3.10 is, and also how difficult it is to generate v3.10.  As noted in https://github.com/CyberBoardPBEM/cbwindows/pull/72#issuecomment-1030707216, using v3.90 is (for now) only slightly limiting, and we already have the code.  If v3.10 is the same as v3.90, then it's almost trivial to provide a switch to save v3.10.

And speaking of getting people to test CB 3.5/4.0:  How can users who aren't accustomed to using Visual Studio able to get pre-built binaries?

wsu-cb commented:
> > The point of the switch is that uptake of the new version has been slow. I hope more people would be willing to run CB 3.5/4.0 Alpha/Beta/Release if they could still interoperate with other users who were still using CB 3.1.
> 
> That's a good point. I guess I hadn't considered 3rd party testing in recent days since earlier builds never seem to be consumed--although I can't be sure about that.
> 
> If we can still produce 3.10 compliant files then that should be the default output version (reverse of what I said earlier.) A 4.0 format switch should be held close to developers until new features are ready that can't be saved in a 3.10 format. The code would no longer output v3.50 format (which looks to be just 3.10 anyway). It would just load it. One benefit and inducement for testing would availability of be the 64bit version of the programs--assuming it can write 3.10 as well.
> 
> > And speaking of getting people to test CB 3.5/4.0: How can users who aren't accustomed to using Visual Studio able to get pre-built binaries?
> 
> I had been building them and putting them in the release area https://github.com/CyberBoardPBEM/cbwindows/releases but haven't done it for quite a while since we've been making big changes to the code. This can restart if we have something stable to make available.
> 
> -Dale

I agree that right now it makes sense to hold off on publishing builds until this PR looks stable.

To summarize:
1)  Tag v3.5-prerelease-2 has the PROGRAM version set to 3.50, and the FILE version set to 3.90.  (To make things even more confusing, the COMMENTS say the FILE version 4.00 is where XtremeToolkit is removed.  In commit 0e66b, I changed the comments to say XtremeToolkit was removed at v3.90 and v4.00 is 32bit ID support.)
2)  The current theory is that FILE v3.90 format is actually the same as FILE v3.10 format; the only difference is the version number.  If at any point we discover this theory is false, this plan goes in the trash can, and we create a new plan.
3)  Previously, I had added a switch /id32cb64 that changed the FILE version from v3.90 to v4.00 (with actual differences in the format).
4)  The plan is to now change the FILE version to default to v3.10.  We will also (continue to) provide switch /id32cb64 to set CB to generate FILE version 4.00 instead of the default v3.10.  This change will be added to this PR.
5)  Since I will be pushing more changes to this PR anyway, I will replace LONG as mentioned in https://github.com/CyberBoardPBEM/cbwindows/pull/72#discussion_r780683007.
6)  Test this PR against the library of games and gameboxes.
7)  Fix bugs discovered by testing.  Those bug fixes will also be performed in this PR.
8)  Once this PR looks reasonably good, merge it to master branch.
9)  Build and publish binaries of the master branch.
10)  Try to get people to actually use the new binaries.

DLLarson commented:
>Does this match your current thinking?

Perfectly! Nicely articulated!

Regarding the program version. It's apart from the file format stuff and is really only used to populate the about box so I thing we're safe to make that something sane irrespective of the file format versioning.

-Dale

wsu-cb commented:
Did the setup of the machine doing the build check change?  The build failure log looks like it thinks MFC isn't installed.

DLLarson commented:
Holy crap! I did get a note from GitHub about something changing in their build stuff. I guess it busted the build.

I'll forward you the Email so you see what I'm talking about.

-Dale

wsu-cb commented:
I pushed an update to this PR that includes your change to the workflow, and it built successfully.  I have also attempted to add x64 builds to the script, and I think the logs tell me that I did so successfully, and that the builds work.  Please see if you interpret the results the same way.

DLLarson commented:
I saw the new build and it looks good to me. All this GitHub action stuff is pretty powerful. You can create installers and tons of other stuff too. All run on their servers.

-Dale

DLLarson reviewed (approved):
Done checking code changes. Next up... verify file format compatibility.

-Dale

wsu-cb commented:
I discovered that, when changing the file version back to 3.10, I missed changing an `ASSERT` that failed when saving files.  (I had been focused on loading files.)  I have fixed the `ASSERT` and `commit`ted it as a `--fixup`.  However, I don't want to do a surprise `push --force` since you have already started looking at the previous version of the PR.  Please let me know whether you prefer to beautify the history, or leave it in its current state.

DLLarson commented:
>However, I don't want to do a surprise push --force since you have already started looking at the previous version of the PR. Please let me know whether you prefer to beautify the history, or leave it in its current state.

Go right ahead. It's not a big deal for me refresh to the latest and rebuild.

-Dale

DLLarson commented:
Hi Bill,

I've tried a variety of gbx and gsn/gam files to test compatibility new saves and the currently released version of CB and haven't hit any issues at all. 

My attempt at automating this sort of this hasn't gone well so far (I have a request for assistance on the program's GitHub site but it just doesn't seem that active overall.) I'm ready to merge this PR if you don't have any other issues on your side. Let me know.

The oldest huge game I could test (World in Flames-aka WiF) really cinched it for me. It had no issues that didn't also show up in the currently shipping CB like deleted pieces for one of the scenarios. Both behaved identically.

-Dale

wsu-cb commented:
Whew!  I'm glad you didn't find any new problems.

As such, I have no additional work to add here.  Please merge at your convenience.
